### ISSUE TITLE
Create bioconda package

### ISSUE THREAD

**Issue by mbhall88** [rmcolq/pandora#244]
There are two things that need to be done before I can start working on this:

- [ ] Merging the current `dev` branch onto `master` (after #243 is merged into `dev`)
- [ ] Tag a release after the above merge. (Bioconda strongly recommend basing recipes on tagged releases)

**Comment by leoisl** [comment]
Hello,

I have a favour to ask you: I have been always postponing improving my knowledge on infrastructure tasks like this. For such things, I think actually doing it is the best way to learn. I am wondering if it would be fine for you to let me do this, and then I do a draft PR, with a few review requests from you. I will read the official docs, but I think you have been creating these recipes for some time now, and might have some tricks and best practices of your own, so I would enjoy also to take a look at some 2 or 3 of your recipes. OFC, feel free to decide that you want to do this yourself, then I can see your recipe later.

**Comment by mbhall88** [comment]
Of course; be my guest. This is probably a good recipe to get thrown in the deep end with too :sweat_smile: Happy to pair-program in the beginning too to get you "up to speed" if you want?

**Comment by leoisl** [comment]
Oh, pair programming in the beginning would be amazing! I am taking more care of my sleeping schedule now though, so doing it 11 PM Cam / 9 AM Sunshine is a bit hard to me. It would be better for me 10 PM Cam / 8 AM Sunshine or 8 AM Cam / 6 PM Sunshine. Any time past 8 AM Cam is also fine. Is any of these time slots ok for you?

**Comment by mbhall88** [comment]
I have opened a PR (https://github.com/bioconda/bioconda-recipes/pull/27388) and asked for help as I am having issues with (surprise) Boost.

**Comment by mbhall88** [comment]
So annoyingly it seems like Hunter actually makes it *harder* to make a bioconda recipe. Maybe you could take a look at this when you get a chance @leoisl as you're the hunter expert here.

**Comment by leoisl** [comment]
Will take a look, this is surprising, I haven't bumped into this issue before.

Trivial question, I think the answer is negative, but it is worth trying anyway: could the precompiled binary make the conda recipe easier? The only requirement to run it is OpenMP 4.0+, which comes with GCC 4.9+. So we would install this GCC version or a later one, and just download the binary and make it executable. Or should we necessarily compile from source? I guess a downside is that the binary just works on Linux, and not on Mac OS, but I have no idea if the current code compiles and runs fine on Mac OS (we don't have Mac OS in our CI, so maybe fine to ignore it)?

**Comment by mbhall88** [comment]
> could the precompiled binary make the conda recipe easier? 

No. It needs to be built from source. I can't find anything in the docs that explicitly states this, but I suspect it is to do the fact that conda uses some very customer build systems with custom compilers and this might cause some problems? I dunno.

**Comment by leoisl** [comment]
I am working on this, and will log some issues here. Bioconda does not like the `-` character in software version, so when we say the version of `pandora` is `0.9.0-rc1`, it errors out with:

```
Error: bad character '-' in package/version: 0.9.0-rc1
```

I changed the bioconda version to be instead `0.9.0_rc1`, but the github version is still `0.9.0-rc1`. Is it fine to leave like this, or do we also change the github version to `0.9.0_rc1`?

**Comment by mbhall88** [comment]
I don't know if a conda package for the RC is a good idea? I'm not even certain if pre-releases are fully supported by conda yet...

I think we just get the recipe working for v0.8.0 first and then worry about trying to do it for the RC after.

**Comment by leoisl** [comment]
Hello, after some 3 days of work exclusively on this, I managed to build pandora on a conda local container (what is called [the bootstrap method](https://bioconda.github.io/contributor/building-locally.html#using-the-bootstrap-method)). This is the PR associated with this work: https://github.com/rmcolq/pandora/pull/275 . The bootstrap method clones the docker containers in conda's Circle CI to build the conda package locally. In other words, if it works with the bootstrap method, which does, it should work on their Circle CI build. We do manage to compile `pandora` on conda's Circle CI, but [25 out of 795 tests fail](https://app.circleci.com/pipelines/github/bioconda/bioconda-recipes/44989/workflows/ffcabc73-bf24-4e47-b8da-6909ec206c82/jobs/151054), all related with denovo. I think the issue might be the files that `GATB` creates, but `GATB` does not create any special files in `temp` or anything, it just uses the workdir.

Very simple tests like this fails:
```
TEST(QueryAbundance, oneKmer_ReturnOne)
{
    Graph graph = Graph::create(new BankStrings("AATGT", NULL),
        "-kmer-size 5 -abundance-min 1 -verbose 0 -out gatb_graph_test");
    auto kmer = "AATGT";
    auto node { graph.buildNode(kmer) };
    const auto covg { graph.queryAbundance(node) };

    // We get the neighbors of this real node and make sure it has the neighbours we
    // expect
    EXPECT_EQ(covg, 1);
    // remove_graph_file("gatb_graph_test");
}
```

As an example, these are the files that `GATB` creates in this test (nothing out of ordinary or special, just files in the workdir):

```
close(3</sys/devices/system/cpu>)       = 0
close(3</home/leandro/git/pandora/cmake-build-debug-coverage>) = 0
close(3</proc/cpuinfo>)                 = 0
close(4</proc/cpuinfo>)                 = 0
close(4</proc/cpuinfo>)                 = 0
close(4</proc/cpuinfo>)                 = 0
close(4</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.0>) = 0
close(5</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.1>) = 0
close(6</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.2>) = 0
close(7</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.3>) = 0
close(8</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.4>) = 0
close(9</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.5>) = 0
close(10</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.6>) = 0
close(11</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_superK_partitions/superKparts.7>) = 0
close(4</proc/cpuinfo>)                 = 0
close(4</proc/cpuinfo>)                 = 0
close(4</proc/cpuinfo>)                 = 0
close(4</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_cfp>) = 0
close(4</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_cfp>) = 0
close(6</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_cfp>) = 0
close(5</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_debloom_partitions.h5 (deleted)>) = 0
close(4</home/leandro/git/pandora/cmake-build-debug-coverage/trashme_43805_cfp (deleted)>) = 0
close(4</proc/cpuinfo>)                 = 0
```

I have absolutely no idea how to solve this

**Comment by leoisl** [comment]
Wrote an issue to their github, will wait for their answer

**Comment by mbhall88** [comment]
> Wrote an issue to their github, will wait for their answer

Can you please link to the issue you raised?

**Comment by leoisl** [comment]
Oops, yeah, sorry: https://github.com/bioconda/bioconda-recipes/issues/28102

**Comment by leoisl** [comment]
Removing the tests worked fine for linux, but failed for mac os: https://github.com/bioconda/bioconda-recipes/pull/27768/checks?check_run_id=2396324449 . Will debug the mac os build and make it work. Then we will have executable where tests don't pass, I will run it on the 4-way pipeline and check if results look similar. If not, then I don't think we should make a conda version until we can make sure the tests pass on the executable

**Comment by leoisl** [comment]
I removed the tests, so linux build works, but I am stuck on the Mac OS build now. Basically, `clang-11` (which we have to forcibly use in the conda build environment, we can't specify another compiler or version) errors out when building `Boost` with this: 

```
clang-11: error: unknown argument: '-fcoalesce-templates'
```

[See full log here](https://app.circleci.com/pipelines/github/bioconda/bioconda-recipes/45275/workflows/fb27d766-693a-4ac8-818d-ebc3d4a25470/jobs/151717).

Basically, this argument does not exist for newer versions of `clang`, because for whatever reason they did not want to preserve backwards compatibility. `Boost` has a check for this (i.e. it just uses this argument if `clang` version is `< 4.0.0`, but the fix has a bug, it makes a lexicographical comparison, and in this case, `11.0.0 < 4.0.0`, and then it tries to use this option that does not exist in `clang 11`. This bug and its fix is [described here](https://github.com/boostorg/build/issues/440). It was fixed on `Boost 1.73`, but for `pandora`, we need to forcibly use `Boost 1.62.0`, so the fix is not yet there. As we also install `Boost` using `Hunter`, we can't add the patch ourselves for the fix... So we have a bug, we know how to fix, but we are using inflexible tools. On one hand, bioconda build environment just accepts one compiler and at a fixed version, and on the other hand, we can't apply patches to `Hunter` builds.

Will try to fix this in the next days, but will take a couple of days of break of trying to build conda recipe

**Comment by leoisl** [comment]
Log for self, this might fix it: https://github.com/alisw/alidist/pull/2159/files

**Comment by leoisl** [comment]
So, I simplified even more stuff. I am skipping the recipe build for `osx`, i.e. we will just have conda recipe for `linux-64`. I don't even know if `pandora` compiles from source and runs fine in `osx` anyway. I know that right now, if `pandora` even works fine on `osx`, it is single threaded (it seems we need to do some more stuff for `OpenMP` to work on `osx`). I also removed the tests, and the `linux-64` recipe now builds: https://github.com/bioconda/bioconda-recipes/pull/27768#issuecomment-833665968 . I will test if this recipe is running fine, as the tests not all pass.

A side question: do we proceed, or do we necessarily need to support `osx`? I personally don't think is necessary `osx` support, as I view `pandora` as a tool to be run on clusters, which are basically `linux` in bioinformatics.

**Comment by mbhall88** [comment]
It's ok for now, but we do need to figure out how to get it to run on Mac. As an example of why this is important, I am developing a [tool](https://github.com/mbhall88/drprg) to do AMR prediction using `pandora`. Some of the people likely to want to run this will most certainly be using Mac. So my tool is currently limited to use on Linux.

**Comment by iqbal-lab** [comment]
Agree with both. In particular,  I think initially supporting Linux only is fine/acceptable

**Comment by leoisl** [comment]
In the worst case, at least for now, your tool can still be run on Mac, as we have containers for `pandora`, although it would be a bit messy... instead of calling `pandora` directly, we can execute it through the singularity container. Far from the ideal solution, and would require Mac users to have `singularity` installed, but it seems fine for a temporary solution...

**Comment by mbhall88** [comment]
[Singularity is not (readily) available on Mac](https://sylabs.io/guides/3.7/admin-guide/installation.html#installation-on-windows-or-mac).

**Comment by leoisl** [comment]
If it was readily available, would it be an option for you? Or even if we have singularity working on osx just as fine it works on linux, would we still need `pandora` to be osx-compatible? Being compatible with osx can be really a pain sometimes, I am just wondering if there is any acceptable way around it...

**Comment by leoisl** [comment]
The 25 tests that fail are now [removed manuallly](https://github.com/bioconda/bioconda-recipes/pull/27768/commits/b50c59fa9f523892dc6d68d2c1f03a6cc74e80bd) in the build script when testing `pandora`. The remaining 770 tests still run and all work. If we can skip `0.8.0`, then I will also test the [example workflow](https://github.com/rmcolq/pandora/tree/0.9.0_branch/example) after these 770 tests, as it is a way to test `pandora discover` also. If all these tests pass, I think we can say all is fine with the binary. I hope they accept the recipe with these tests, as it literally takes 1.5 seconds to test:

> 22:55:44 BIOCONDA INFO (OUT) [==========] 770 tests from 114 test suites ran. (1503 ms total)

.... while it takes 1h to build...

**Comment by mbhall88** [comment]
>  Is it fine then to skip 0.8.0 and push a recipe only for 0.9.0?

Sure thing

**Comment by mbhall88** [comment]
Done in https://github.com/bioconda/bioconda-recipes/pull/27768
